src/net.h

```diff
@@ -724,4 +682,158 @@ class CAddrDB
 };
 
+//// This definition can be moved into CConnman after boost::thread refactoring
+struct ListenSocket {
+    SOCKET socket;
+    bool whitelisted;
+
+    ListenSocket(SOCKET socket, bool whitelisted) : socket(socket), whitelisted(whitelisted) {}
+};
+
+/** Used to pass flags to the Bind() function */
+enum BindFlags {
+    BF_NONE         = 0,
+    BF_EXPLICIT     = (1U << 0),
+    BF_REPORT_ERROR = (1U << 1),
+    BF_WHITELIST    = (1U << 2),
+};
+
+namespace zen {
+typedef struct _NODE_ADDR {
+    std::string ipAddr;
+    int64_t time; // time in msec, of an attempt to connect via TLS
+
+    _NODE_ADDR(std::string _ipAddr, int64_t _time = 0) : ipAddr(_ipAddr), time(_time) {}
+    bool operator==(const _NODE_ADDR b) const
+    {
+        return (ipAddr == b.ipAddr);
+    }
+} NODE_ADDR, *PNODE_ADDR;
+}
+
+class CConnman {
+public:
+    struct Options
+    {
+        uint64_t nLocalServices = 0;
+        int nMaxConnections = 0;
+        unsigned int nSendBufferMaxSize = 0;
+        unsigned int nReceiveFloodSize = 0;
+        
+        std::vector<CSubNet> vWhitelistedRange;
+    };
+
+    void Init(const Options& connOptions)
+    {
+        nLocalServices = connOptions.nLocalServices;
+        nMaxConnections = connOptions.nMaxConnections;
+        nSendBufferMaxSize = connOptions.nSendBufferMaxSize;
+        nReceiveFloodSize = connOptions.nReceiveFloodSize;
+        {
+            LOCK(cs_vWhitelistedRange);
+            vWhitelistedRange = connOptions.vWhitelistedRange;
+        }
+    }
+
+    void StartNode(boost::thread_group& threadGroup, CScheduler& scheduler , const Options& connOptions);
+    bool StopNode();
+    void Stop();
+    void NetCleanup();
+    void Interrupt();
+
+    bool Bind(const CService &addr, unsigned int flags);
+    bool BindListenPort(const CService &bindAddr, std::string& strError, bool fWhitelisted = false);
+
+    void AddOneShot(const std::string& strDest);
+    void ProcessOneShot();
+
+    bool OpenNetworkConnection(const CAddress& addrConnect, CSemaphoreGrant *grantOutbound = NULL, const char *strDest = NULL, bool fOneShot = false);
+    void AcceptConnection(const ListenSocket& hListenSocket);
+    CNode* FindNode(const CNetAddr& ip);
+    CNode* FindNode(const CSubNet& subNet);
+    CNode* FindNode(const std::string& addrName);
+    CNode* FindNode(const CService& ip);
+    CNode* ConnectNode(CAddress addrConnect, const char *pszDest = NULL);
+    bool IsWhitelistedRange(const CNetAddr &ip);
+    bool AttemptToEvictConnection(bool fPreferNewConnection);
+
+    void SocketSendData(CNode *pnode);
+
+    CConnman();
+    ~CConnman();
+
+    bool fAddressesInitialized {false};
+    std::unique_ptr<CSemaphore> semOutbound = nullptr;
+    std::unique_ptr<CNode> pnodeLocalHost = nullptr;
+
+    std::vector<CNode*> vNodes;
+    CCriticalSection cs_vNodes;
+    std::vector<std::string> vAddedNodes;
+    CCriticalSection cs_vAddedNodes;
+    std::list<CNode*> vNodesDisconnected;
```

Comment:
I see `vNodesDisconnected` is used in just two places, `CConnman::NetCleanup` and `ThreadSocketHandler`; actually it seems to me there is no real synchronization of those two, since `ThreadSocketHandler` runs indefinitely and `CConnman::NetCleanup` is called in `Shtudown` and in `CConnman` destructor. Is it worth to add a locking mechanism through `CCriticalSection`?

Reply:
Ok makes sens, thanks for the explanation.